File: httpserver/Resource.hs
module Resource (
	Resource(..),
	StreamResult(..),
	BodyStream,
	handleRequest) where

import Method 
import qualified Request as Req
import qualified Response as Resp
import qualified Data.ByteString.Char8 as B8
import qualified Data.ByteString as BS
import DataDecl


data Resource = Resource { 
	path :: String,
	method :: Method,
	callback :: Req.Request -> BodyStream -> IO Resp.Response
}

instance Show Resource where
	show r = "Resource {path=" ++ show (path r) ++", method=" ++ show (method r) ++ "}"


handleRequest :: Resource -> Req.Request -> BodyStream -> IO Resp.Response
handleRequest = callback

File: rust_http_server/resource.rs
use http::method::Method;
use http::request::Request; 
use http::response::Response;

//type RequestCallback<'a> = &'a Fn((Request, Response));

#[derive(Show)]
pub struct Resource<'a, F> where F: Fn(&'a mut Request, &'a mut Response) -> () {
	path: String,
	method: Method,
	request_callback: F //RequestCallback,
}

impl<'a, F> Resource<'a, F> where F: Fn(&'a mut Request, &'a mut Response) -> () {
	pub fn init<'r>(path: &str, method: Method, callback: F) -> Resource<'r, F> {
		Resource { path: path.to_string(), method: method, request_callback: callback}
	}
}